### ISSUE TITLE
Rule proposal: `no-unused-return-value`

### ISSUE THREAD

**Issue by silverwind** [eslint/eslint#12980]
**Please describe what the rule should do:**

Detect unused return values of built-in object methods  where it never makes sense *not* to use the return value.

**What category of rule is this? (place an "X" next to just one item)**

[x] Suggests an alternate way of doing something (suggestion)

**Provide 2-3 code examples that this rule will warn about:**

```js
const arr = [{name: 'eslint'}];
arr.map(item => item.name); // error: Unused Array.prototype.map return value
```

```js
const arr = [{name: 'eslint'}];
arr.filter(item => item.name === 'eslint'); // error: Unused Array.prototype.filter return value
```
```js
let i = 0;
const arr = [1,2,3];
arr.map((item) => {  // error: Unused Array.prototype.map return value
  i += 1;
});
```
```js
Object.create(null); // error: Unused Object.create return value
```

**Why should this rule be included in ESLint (instead of a plugin)?**

It is useful to anyone using built-in JS objects and especially for iteration methods that return a value like `Array.prototype.map` . If its return value is unused, it would be better to refactor such cases to `Array.prototype.forEach` to make the intend clear.

I could see the rule being especially useful to beginners who might not have memorized yet that `Array.prototype.filter` or `Array.prototype.map` do not modify the array.

The concept can be applied to any built-in static or prototype method that either returns a value or does iteration only. It can be extended to other built-ins like `String`, `Number`, `Map`, `Set` and more.

For `Array` and `Object`, I see these suitable methods:
```
Array.isArray
Array.of
Array.prototype.concat
Array.prototype.filter
Array.prototype.includes
Array.prototype.indexOf
Array.prototype.join
Array.prototype.lastIndexOf
Array.prototype.slice
Array.prototype.toSource
Array.prototype.toString
Array.prototype.toLocaleString
Array.prototype.entries
Array.prototype.every
Array.prototype.find
Array.prototype.findIndex
Array.prototype.keys
Array.prototype.map
Array.prototype.some
Array.prototype.values

Object.create
Object.entries
Object.fromEntries
Object.getOwnPropertyDescriptor
Object.getOwnPropertyDescriptors
Object.getOwnPropertyNames
Object.getOwnPropertySymbols
Object.getPrototypeOf
Object.is
Object.isExtensible
Object.isFrozen
Object.isSealed
Object.keys
Object.values
Object.prototype.hasOwnProperty
Object.prototype.isPrototypeOf
Object.prototype.propertyIsEnumerable
Object.prototype.toLocaleString
Object.prototype.toString
```

Related previous discussion: https://github.com/eslint/eslint/issues/12269

**Are you willing to submit a pull request to implement this rule?**

No

**Comment by ljharb** [comment]
What happens when I’m using .map on an arbitrary object that doesn’t provide a forEach method?

**Comment by kaicataldo** [comment]
My biggest concern with this proposal is that in the cases where we're dealing with methods on the prototype we'll be playing a guessing game as to what type of object we're dealing with, though there is precedence for this with [array-callback-return](https://eslint.org/docs/rules/array-callback-return#top). However the other methods do seem enforceable without false positives.

**Comment by ljharb** [comment]
`Array.from` is a very important method to use in an environment without native iteration syntax, and often the return value is not needed.

**Comment by silverwind** [comment]
> What happens when I’m using .map on an arbitrary object that doesn’t provide a forEach method?

> we'll be playing a guessing game as to what type of object we're dealing with

Yeah, the fact that not all types can be statically determined would certainly dampen the rule's usefulness. It should work considerably better in a TypeScript environment.

> Array.from is a very important method to use in an environment without native iteration syntax, and often the return value is not needed.

Care to give an example? Something like `Array.from(arrayLike).forEach()` still counts as using the return value, or do you mean something entirely different?

**Comment by ljharb** [comment]
`Array.from(iterableOrArraylike, forEachFunction)`

**Comment by silverwind** [comment]
Right, I wasn't aware of those `Array.from` signatures, removing it form the list.

**Comment by ljharb** [comment]
Additionally, `Object.entries` and `Object.values` can trigger side effects from getters, and `Object.prototype.toSource` isn't part of the language or a thing outside of Firefox.

Additionally, my polyfills/shims often use try/catches around no-op statements with builtin methods (including these) to feature-detect behavior. Rollup for awhile had a flawed heuristic where it tried to delete "unused code" - like this rule would warn on - and it broke airbnb.com in IE 11.

**Comment by silverwind** [comment]
`Object.prototype.toSource` removed. I'm not sure if getter side effects really justify removal. I guess more then just `Object.entries` annd `Object.values` could trigger getters and if/when they trigger might even be implementation-specific behaviour.

**Comment by ljharb** [comment]
It wouldn’t be implementation specific; it’s that user code could be relying on it.

**Comment by silverwind** [comment]
You're right, getters should not be disregarded. Still, I think if the goal is to explicitely trigger getters, there are other ways to achieve the same (e.g. regular `for` loop for example).

Overall, I think the usefulness of this rule depends heavily on how reliable `eslint`'s type detection works.
